### PR TITLE
Make Push and Pull Replication follow builder pattern.

### PR DESCRIPTION
**Pull request by rhyshort**
Make Push and Pull replication follow builder pattern by creating
APIs so you can do the following:

```
new PushReplication().source().target().start()

```

Split out from PR #181 

reviewer @mikerhodes 
reviewer @tomblench

### CODE REVIEW

File: sync-core/src/main/java/com/cloudant/sync/replication/Replication.java [hunk 1]
```diff
@@ -154,6 +153,16 @@ public String toQueryString() {
     protected Replication() {
         /* prevent instances of this class being constructed */
     }
+    /**
```
Comment (tomblench) [inline sync-core/src/main/java/com/cloudant/sync/replication/Replication.java]: Add newline above

File: sync-core/src/main/java/com/cloudant/sync/replication/PullReplication.java [hunk 2]
```diff
@@ -46,6 +46,8 @@
      */
     public Filter filter;
 
+
```
Comment (tomblench) [inline sync-core/src/main/java/com/cloudant/sync/replication/PullReplication.java]: Delete these extra newlines

File: sync-core/src/main/java/com/cloudant/sync/replication/PullReplication.java [hunk 3]
```diff
@@ -54,6 +56,28 @@ public PullReplication() {
         /* Does nothing but we can now document it */
     }
 
+
```
Comment (tomblench) [inline sync-core/src/main/java/com/cloudant/sync/replication/PullReplication.java]: Delete these extra newlines - we should normally have one blank line between methods

**Comment by tomblench** [comment]
+1 once you've fixed up newlines

File: sync-core/src/main/java/com/cloudant/sync/replication/PullReplication.java (hunk 3 above)
Comment (rhyshort) [inline sync-core/src/main/java/com/cloudant/sync/replication/PullReplication.java]: will do, most of these were probably caused as the split commits.

**Comment by mikerhodes** [comment]
For now I'd rather this stuck with a `build()` rather than `start()` method. There's no need to change that API right now I don't think.

**Comment by rhyshort** [comment]
Okay, changed `start` to `build` in 62549bf and removed the line which started the replicator

File: sync-core/src/main/java/com/cloudant/sync/replication/Filter.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+     *                   constructing the {@code _changes} feed call for the remote database.
+     *                   Integer values should be added as String objects.
+     *
+     * @see <a href="http://docs.couchdb.org/en/1.6.x/couchapp/ddocs.html#filter-functions">Filter functions CouchDB docs</a>
+     */
+    public Filter(String filterName, Map<String, String> parameters) {
+        this.name = filterName;
+        ImmutableMap.Builder<String, String> builder = new ImmutableMap.Builder<String, String>();
+        builder.putAll(parameters);
+        this.parameters = builder.build();
```
Comment (mikerhodes) [inline sync-core/src/main/java/com/cloudant/sync/replication/Filter.java]: We can do this without using Guava (presuming it's fine on Android):

```
this.parameters = Collections.unmodifiableMap(new Hashmap<>(parameters));
```
http://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#unmodifiableMap%28java.util.Map%29

File: sync-core/src/main/java/com/cloudant/sync/replication/Filter.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+ * Map&lt;String, String&gt; params = new HashMap();
+ * map.put("max_age", "10");
+ * map.put("name", "john");
+ * Filter filter = new Filter("doc/filterName", map);
+ * </pre>
+ *
+ * @see <a href="http://couchdb.readthedocs.org/en/1.6.x/replication/intro.html#controlling-which-documents-to-replicate">Controlling documents replicated</a>
+ * @see <a href="http://docs.couchdb.org/en/1.6.x/couchapp/ddocs.html#filter-functions">Filter functions CouchDB docs</a>
+ */
+public class Filter {
```
Comment (mikerhodes) [inline sync-core/src/main/java/com/cloudant/sync/replication/Filter.java]: I think this is PullReplication specific, should we name it such?

File: sync-core/src/main/java/com/cloudant/sync/replication/ReplicatorBuilder.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+            replication = pullReplication;
+
+
+        } else if (pullTarget == null && pullSource == null){
+            //should be a push
+            if(pushSource == null || pushTarget == null){
+                //configured wrong.
+                throw new IllegalStateException("Push replication not correctly configured, " +
+                        "target or source is null");
+            }
```
Comment (mikerhodes) [inline sync-core/src/main/java/com/cloudant/sync/replication/ReplicatorBuilder.java]: Validate no filter applied.

File: sync-core/src/main/java/com/cloudant/sync/replication/Filter.java (hunk 5 above)
Comment (rhyshort) [inline sync-core/src/main/java/com/cloudant/sync/replication/Filter.java]: Sure, we can do that

File: sync-core/src/main/java/com/cloudant/sync/replication/Filter.java (hunk 4 above)
Comment (rhyshort) [inline sync-core/src/main/java/com/cloudant/sync/replication/Filter.java]: Yeah we can do that in android http://developer.android.com/reference/java/util/Collections.html#unmodifiableMap(java.util.Map<?%20extends%20K,%20?%20extends%20V>)

File: sync-core/src/main/java/com/cloudant/sync/replication/ReplicatorBuilder.java [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+    /**
+     * Sets the source database for a pull replication
+     * @param pullSource The source database for a pull replication
+     * @return This instance of {@link ReplicatorBuilder}
+     */
+    public E from(S pullSource) {
+        this.source = pullSource;
+        return (E) this;
+    }
```
Comment (mikerhodes) [inline sync-core/src/main/java/com/cloudant/sync/replication/ReplicatorBuilder.java]: Correct the java doc to remove the pull-specific-ness.

File: sync-core/src/main/java/com/cloudant/sync/replication/Filter.java (hunk 5 above)
Comment (rhyshort) [inline sync-core/src/main/java/com/cloudant/sync/replication/Filter.java]: done in 22f9fcf

File: sync-core/src/main/java/com/cloudant/sync/replication/Filter.java (hunk 4 above)
Comment (rhyshort) [inline sync-core/src/main/java/com/cloudant/sync/replication/Filter.java]: done in 22f9fcf

File: sync-core/src/main/java/com/cloudant/sync/replication/ReplicatorBuilder.java [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+     * Creates a pull replication builder.
+     * @return A newly created {@link Pull} replication builder
+     */
+    public static Pull pull(){
+        return new Pull();
+    }
+
+    /**
+     * Creates a {@link Push} replication builder.
+     * @return
```
Comment (tomblench) [inline sync-core/src/main/java/com/cloudant/sync/replication/ReplicatorBuilder.java]: I get a javadoc warning here: `warning: no description for @return`

File: sync-core/src/main/java/com/cloudant/sync/replication/Filter.java (hunk 4 above)
Comment (mikerhodes) [inline sync-core/src/main/java/com/cloudant/sync/replication/Filter.java]: Would be good to make that change here.

File: sync-core/src/main/java/com/cloudant/sync/replication/ReplicatorBuilder.java (hunk 7 above)
Comment (rhyshort) [inline sync-core/src/main/java/com/cloudant/sync/replication/ReplicatorBuilder.java]: done in 4205568

File: sync-core/src/main/java/com/cloudant/sync/replication/Filter.java (hunk 5 above)
Comment (mikerhodes) [inline sync-core/src/main/java/com/cloudant/sync/replication/Filter.java]: `PullFilter` so it's not too huge a name? It's in the replication package so it's clear that's it replication specific.

File: sync-core/src/main/java/com/cloudant/sync/replication/ReplicatorBuilder.java (hunk 8 above)
Comment (rhyshort) [inline sync-core/src/main/java/com/cloudant/sync/replication/ReplicatorBuilder.java]: fixed in 796de43

**Comment by mikerhodes** [comment]
+1, looks great.

**Comment by tomblench** [comment]
Do we do CHANGES and doc updates here or elsewhere?

**Comment by rhyshort** [comment]
I plan to do that in another PR after the Filter API has been exposed

**Pull request merged**